feat(ws): Notebooks 2.0 // Frontend // Workspaces details // Pod template tab #173

liavweiss opened the pull request:
As requested I added a code editor for the Pod Template section under the workspace details.
![image](https://github.com/user-attachments/assets/eb7cb395-e757-40a8-8244-6ab88da52264)


The editor will display the all yaml (mocked in the code) as read-only and will utilize the full height of its parent.

paulovmr commented on workspaces/frontend/src/app/pages/Workspaces/Details/WorkspaceDetailsPodTemplate.tsx:
```diff
@@ -0,0 +1,40 @@
+      isLineNumbersVisible
+      isFullHeight
+      isReadOnly
+      code={podTemplateYaml || '# No podTemplate data available'}
```
```suggestion
      code={podTemplateYaml || '# No pod template data available'}
```

paulovmr commented on workspaces/frontend/src/app/pages/Workspaces/Details/WorkspaceDetails.tsx:
```diff
@@ -72,7 +73,9 @@ export const WorkspaceDetails: React.FunctionComponent<WorkspaceDetailsProps> =
           >
             <TabContent id="podTemplateBodyPadding">
-              <TabContentBody hasPadding>Pod template</TabContentBody>
+              <TabContentBody hasPadding style={{ height: '500px' }}>
```
Ideally, the tab content body should fill the remaining height of the drawer until the bottom of the drawer, while keeping the scroll in the code editor.

paulovmr reviewed (changes requested):
Thanks for the PR, @liavweiss !

jenny-s51 commented on workspaces/frontend/src/app/pages/Workspaces/Details/WorkspaceDetailsPodTemplate.tsx:
```diff
@@ -0,0 +1,40 @@
+import React from 'react';
```
```suggestion
import React from 'react';
import { CodeEditor, CodeEditorProps, Language } from '@patternfly/react-code-editor';
import yaml from 'js-yaml';

const mockYaml = `apiVersion: kubeflow.org/v1beta1
kind: Workspace
metadata:
  name: jupyterlab-workspace
spec:
  paused: false
  deferUpdates: false
  kind: "jupyterlab"
  podTemplate:
    podMetadata:
      labels: {}
      annotations: {}
    volumes:
      home: "workspace-home-pvc"
      data:
        - pvcName: "workspace-data-pvc"
          mountPath: "/data/my-data"
          readOnly: false
    options:
      imageConfig: "jupyterlab_scipy_190"
      podConfig: "tiny_cpu"`;

type WorkspaceDetailsPodTemplateProps = Omit<CodeEditorProps, 'ref'> & {
  testId?: string;
  codeEditorHeight?: string;
};
const parsedYaml = yaml.load(mockYaml);
const podTemplateYaml = yaml.dump(parsedYaml || {});

export const WorkspaceDetailsPodTemplate: React.FC<Partial<WorkspaceDetailsPodTemplateProps>> = ({
  // 38px is the code editor toolbar height+border
  // calculate the div height to avoid container scrolling
  height = 'calc(100% - 38px)',
  codeEditorHeight = '650px',
  ...props
}) => (
  <div data-testid={props.testId} style={{ height, padding: '14px' }}>
    <CodeEditor
      isLineNumbersVisible
      isReadOnly
      code={podTemplateYaml || '# No pod template data available'}
      height={codeEditorHeight}
      language={Language.yaml}
      {...props}
    />
  </div>
);
```

liavweiss replied:
Thank you Jenny, much appreciate!
I added it to my code and tested it, and now everything works.

jenny-s51 reviewed (changes requested):
@liavweiss great work on this! I've added a workaround below to fix the issue you were running into with the drawer height.

jenny-s51 commented on workspaces/frontend/src/app/pages/Workspaces/Details/WorkspaceDetails.tsx:
```diff
@@ -72,7 +73,9 @@ export const WorkspaceDetails: React.FunctionComponent<WorkspaceDetailsProps> =
           >
             <TabContent id="podTemplateBodyPadding">
-              <TabContentBody hasPadding>Pod template</TabContentBody>
+              <TabContentBody hasPadding style={{ height: '500px' }}>
```
```suggestion
              <TabContentBody hasPadding>
```

Inline style can be removed once the suggested fix is applied.

liavweiss replied:
I removed it, thanks!

liavweiss commented:
After help from Jenny, and fixed all comments
![image](https://github.com/user-attachments/assets/e8418c59-123d-4860-a82e-f1681fc5e2fe)

paulovmr commented on workspaces/frontend/package.json:
```diff
@@ -94,10 +94,13 @@
     "webpack-merge": "^5.10.0"
   },
   "dependencies": {
+    "@patternfly/react-code-editor": "^6.1.0",
```
Probably best to keep it at `6.0.0` atm to align with the other PatternFly dependencies.
```suggestion
    "@patternfly/react-code-editor": "^6.0.0",
```

paulovmr reviewed (changes requested):
Hi @liavweiss  @jenny-s51 , thanks for the PR and changes. I tried it locally, and I get the behavior below. Was the code editor supposed to adjust its height to align with the bottom of the screen?

[codeeditorheight.webm](https://github.com/user-attachments/assets/260cecf2-646f-44e8-bd05-984fa9e4aec0)

jenny-s51 commented on workspaces/frontend/package.json:
```diff
@@ -94,10 +94,13 @@
     "@patternfly/react-core": "^6.0.0",
     "@patternfly/react-icons": "^6.0.0",
     "@patternfly/react-styles": "^6.0.0",
     "@patternfly/react-table": "^6.0.0",
```
```suggestion
    ""@patternfly/react-styles": "^6.0.0",
    "@patternfly/react-table": "^6.0.0",
```

jenny-s51 commented on workspaces/frontend/src/app/pages/Workspaces/Details/WorkspaceDetails.tsx:
```diff
@@ -72,7 +73,9 @@ export const WorkspaceDetails: React.FunctionComponent<WorkspaceDetailsProps> =
           >
             <TabContent id="podTemplateBodyPadding">
-              <TabContentBody hasPadding>Pod template</TabContentBody>
+              <TabContentBody hasPadding>
+                <WorkspaceDetailsPodTemplate />
+              </TabContentBody>
```
```suggestion
              <TabContentBody className="pf-v6-u-h-100" hasPadding>
                <WorkspaceDetailsPodTemplate codeEditorHeight="100%" />
              </TabContentBody>
```

jenny-s51 commented on workspaces/frontend/src/app/pages/Workspaces/Details/WorkspaceDetails.tsx:
```diff
@@ -72,7 +73,9 @@ export const WorkspaceDetails: React.FunctionComponent<WorkspaceDetailsProps> =
             aria-label="Pod template"
           >
             <TabContent id="podTemplateBodyPadding">
```
```suggestion
            <TabContent style={{ flex: 1 }} className="pf-v6-u-h-100" id="podTemplateBodyPadding">

```

liavweiss commented:
I added a fix for all the comments, and I fixed the height=100% issue for the code editor

fullscreen:
![image](https://github.com/user-attachments/assets/8e0923ee-a09a-4be3-83f0-c7f71f80e773)

half screen
![image](https://github.com/user-attachments/assets/034e5e5b-d528-477e-819e-a76883db4a41)

jenny-s51 reviewed (approved):
Tested locally – great work on this, @liavweiss. The code editor behaves as expected and adjusts to take up the full height of the parent container when screen is resized. LGTM

ederign commented:
/ok-to-test

ederign commented:
/lgtm
